### PR TITLE
Object Explorer API

### PR DESCRIPTION
**Pull request by MattIrv**
This PR adds a basic object explorer extensibility API that will allow extensions to navigate active connections in OE and expand, select, and collapse specific nodes, as well as go directly to a node given its node path.

This will allow us to implement features like https://github.com/Microsoft/sqlopsstudio/issues/572

### CODE REVIEW

File: src/sql/sqlops.d.ts [hunk 1]
```diff
@@ -108,6 +108,68 @@ declare module 'sqlops' {
...
+		 * @returns {ObjectExplorerNode} The node corresponding to the given connection and path,
+		 * or undefined if no such node exists.
+		*/
+		export function getNode(connectionId: string, nodePath?: string): Thenable<ObjectExplorerNode>;
+
+		/**
+		 * Get all active Object Explorer connection nodes
+		 * @returns {ObjectExplorerNode[]} The Object Explorer nodes for each saved connection
+		*/
+		export function getActiveConnections(): Thenable<ObjectExplorerNode[]>;
```
Comment (llali) [inline src/sql/sqlops.d.ts]: >getActiveConnections [](start = 18, length = 20)

getActiveConnectionNodes?

Comment (llali) [reply src/sql/sqlops.d.ts]: also the result should be map of the connection id and nodes right? 


---
In reply to: [170717086](https://github.com/Microsoft/sqlopsstudio/pull/783#discussion_r170717086) [](ancestors = 170717086)

File: src/sql/parts/registeredServer/common/objectExplorerService.ts [hunk 2]
```diff
@@ -65,16 +65,29 @@ export interface IObjectExplorerService {
 	onSelectionOrFocusChange: Event<void>;
 
 	getServerTreeView(): ServerTreeView;
+
+	getActiveConnections(): { connectionId: string, nodeInfo: sqlops.NodeInfo }[];
+
+	getChildren(connectionId: string, nodePath: string): Thenable<sqlops.NodeInfo[]>;
```
Comment (kevcunnane) [inline src/sql/parts/registeredServer/common/objectExplorerService.ts]: Have we looked at the built-in [TreeDataProvider APIs](https://code.visualstudio.com/docs/extensionAPI/vscode-api#TreeDataProvider) and considered how we can leverage their patterns / improve on their design? 
- For contributors to the tree, they have a pattern of "getChildren returns a set of objects, and given the objects you can request a TreeItem using getTreeItem".
- In this case it seems NodeInfo is a lot like TreeItem, but it can't be abstracted away depending on the node provider.

Similarly they have a way to provide tri-state expansion (expanded/collapsed/node, where node means it's a leaf node). 

What VSCode seems to lack is programmatic support for "select this node" etc. so it's great we're adding this in. The closer we can align the easier for contributors to use our code & for us to potentially contribute back to VSCode down the line.

File: src/sql/parts/registeredServer/common/objectExplorerService.ts [hunk 3]
```diff
@@ -65,16 +65,29 @@ export interface IObjectExplorerService {
...
+
+	isExpanded(connectionId: string, nodePath: string): boolean;
+
+	expandNodeForConnection(connectionId: string, nodePath: string): Thenable<void>;
```
Comment (kevcunnane) [inline src/sql/parts/registeredServer/common/objectExplorerService.ts]: Are we going to document these APIs now people can use them? In particular it's important to understand a) how to get a connection ID (what object would you pull this from) and b) how to build up a nodePath. Is this a standard thing? Would there be a way to leverage a method to build it for you?

File: src/sql/parts/registeredServer/common/objectExplorerService.ts [hunk 4]
```diff
@@ -65,16 +65,29 @@ export interface IObjectExplorerService {
 	onSelectionOrFocusChange: Event<void>;
 
 	getServerTreeView(): ServerTreeView;
+
+	getActiveConnections(): { connectionId: string, nodeInfo: sqlops.NodeInfo }[];
```
Comment (kevcunnane) [inline src/sql/parts/registeredServer/common/objectExplorerService.ts]: Can you provide an interface for this object return type? It's standard in the VSCode APIs and makes life easier.

File: src/sqltest/parts/connection/objectExplorerService.test.ts [hunk 5]
```diff
@@ -503,4 +513,203 @@ suite('SQL Object Explorer Service tests', () => {
 		let selectedProfileAndDatabase = objectExplorerService.getSelectedProfileAndDatabase();
 		assert.equal(selectedProfileAndDatabase, undefined);
 	});
+
+	test('isExpanded returns true when the node and its parents are expanded', (done) => {
```
Comment (kevcunnane) [inline src/sqltest/parts/connection/objectExplorerService.test.ts]: Great to see unit tests added for the new APIs.

File: src/sql/parts/registeredServer/common/objectExplorerService.ts [hunk 6]
```diff
@@ -65,16 +65,29 @@ export interface IObjectExplorerService {
...
+
+	getChildren(connectionId: string, nodePath: string): Thenable<sqlops.NodeInfo[]>;
+
+	isExpanded(connectionId: string, nodePath: string): boolean;
+
+	expandNodeForConnection(connectionId: string, nodePath: string): Thenable<void>;
+
+	collapseNodeForConnection(connectionId: string, nodePath: string): Thenable<void>;
```
Comment (kevcunnane) [inline src/sql/parts/registeredServer/common/objectExplorerService.ts]: A few thoughts on these methods:
- we've got 3 separate actions already. I'd guess we'll have more. Would a "getNode" action to return the Node class be better, and then hang actions off of this?
- I'd suggest a single method for expand/collapse such as `setExpandedState(TreeItemCollapsibleState.Collapsed)` rather than separate methods
- Similarly, do a `setSelected(true/false)` method since it's also useful to allow unselect or similar.
- Consider how to support multi-selection vs single-selection. VSCode has been moving towards supporting multi-select and we will no doubt need to also.

File: src/sql/parts/registeredServer/common/objectExplorerService.ts [hunk 7]
```diff
@@ -423,4 +445,188 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
+					return Promise.reject('There is no object at the given node path');
+				}
+				return new Promise((resolve, reject) => {
+					// Recursively expand the tree to get to the given node
+					let expandFunction = ((treeNode: TreeNode) => {
+						if ((stopAtParent || treeNode.isAlwaysLeaf) && treeNode.nodePath === nodePath) {
+							resolve();
+							return;
+						} else if (treeNode.isAlwaysLeaf) {
+							reject('Could not find tree node for expand due to leaf node on path');
```
Comment (kevcunnane) [inline src/sql/parts/registeredServer/common/objectExplorerService.ts]: Please localize all messages that might bubble up to the user

File: src/sql/parts/registeredServer/common/objectExplorerService.ts [hunk 8]
```diff
@@ -423,4 +445,188 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
+
+	public expandNodeForConnection(connectionId: string, nodePath: string, stopAtParent: boolean = false): Thenable<void> {
+		let rootNode = this._activeObjectExplorerNodes[connectionId];
+		if (!rootNode) {
+			return Promise.reject('The given connection is not active in Object Explorer');
+		}
+		let node = this.findTreeNode(connectionId, nodePath);
+		if (!node) {
+			// The node is not in the tree yet, so expand its parent nodes to make it visible
+			return this.findNodeInfo(connectionId, nodePath).then(nodeInfo => {
```
Comment (kevcunnane) [inline src/sql/parts/registeredServer/common/objectExplorerService.ts]: Minor: I've been using async/await recently and it simplifies things a lot, especially where we have nested thenables with promises, more thenables, etc. Are we using it inside our ops studio code yet? This function seems like a nice place to try it out.

Comment (kevcunnane) [reply src/sql/parts/registeredServer/common/objectExplorerService.ts]: Also given the logic here, maybe just wrap this in a private method so you get a simpler flow. There's a lot of steps in each sub-clause, usually means we should separate into their own methods.
```
if (!node) {
    return this.findAndSetExpanded(...);
} else { 
    return this.expandToNode(...);
}
```

File: src/sql/parts/registeredServer/viewlet/serverTreeView.ts [hunk 9]
```diff
@@ -428,6 +429,42 @@ export class ServerTreeView {
 		return this._tree.isDOMFocused();
 	}
 
+	/**
+	 * Expand the given element in the tree
+	 */
+	public expand(element: TreeNode | ConnectionProfile): Thenable<void> {
```
Comment (kevcunnane) [inline src/sql/parts/registeredServer/viewlet/serverTreeView.ts]: Again, combining with collapse as `setExpanded(...)` seems cleaner.

File: src/sql/parts/registeredServer/viewlet/serverTreeView.ts [hunk 10]
```diff
@@ -428,6 +429,42 @@ export class ServerTreeView {
...
+	public reveal(element: TreeNode | ConnectionProfile): Thenable<void> {
+		return this._tree.reveal(element);
+	}
+
+	/**
+	 * Select the given element in the tree and clear any other selections
+	 */
+	public select(element: TreeNode | ConnectionProfile): void {
+		this._tree.clearSelection();
+		this._tree.select(element);
```
Comment (kevcunnane) [inline src/sql/parts/registeredServer/viewlet/serverTreeView.ts]: We can have multi-select now, correct? If so can we ensure this method aligns with multi-select patterns inside of VSCode APIs?

File: src/sql/workbench/api/node/sqlExtHost.protocol.ts [hunk 11]
```diff
@@ -304,6 +304,8 @@ export abstract class ExtHostDataProtocolShape {
 	$stopSession(handle: number, sessionId: string): Thenable<boolean> { throw ni(); }
 }
 
+export abstract class ExtHostObjectExplorerShape { }
```
Comment (anthonydresser) [inline src/sql/workbench/api/node/sqlExtHost.protocol.ts]: This is the old way to do this. This should be an interface now.

**Comment by kburtram** [comment]
@alextercete please take a look at this PR to see if there are gaps in what's needed for #572.

File: src/sql/sqlops.d.ts [hunk 12]
```diff
@@ -108,6 +108,68 @@ declare module 'sqlops' {
...
+		/**
+		 * Get an Object Explorer node corresponding to the given connection and path. If no path
+		 * is given, it returns the top-level node for the given connection. If there is no node at
+		 * the given path, it returns undefined.
+		 * @param {string} connectionId The id of the connection that the node exists on
+		 * @param {string?} nodePath The path of the node to get
+		 * @returns {ObjectExplorerNode} The node corresponding to the given connection and path,
+		 * or undefined if no such node exists.
+		*/
+		export function getNode(connectionId: string, nodePath?: string): Thenable<ObjectExplorerNode>;
```
Comment (alextercete) [inline src/sql/sqlops.d.ts]: From an API consumer's point of view, it's not clear how I'm supposed to get a hold of `nodePath`. After some trial and error, I managed to find the right format and use the API successfully, but I've run into a couple of oddities in the process (in this example, the path was `localhost,1401/Databases/WideWorldImporters/Tables/Application.Cities (System-Versioned)`):

1. The first segment of the path needs to be the server name, which is a bit redundant given I'm already providing the `connectionId` -- I'd expect the path to be _relative_ to the connection.
2. I need to know the folder structure of the node I want to access, and the node's path is coupled to the text used for display purposes -- for example, I'd expect the last segment of the path to be `Application.Cities` instead of `Application.Cities (System-Versioned)`. Ideally, I'd like to select an object by specifying its database, schema and object names. However, I understand that might be too high-level, so maybe it'd be nice if there were factory methods (maybe on `ObjectExplorerNodePath`) that allowed building paths to common objects, and then allowing navigation from there:

    ```typescript
    const path = ObjectExplorerNodePath.forTable('WideWorldImporters', 'Application', 'Cities');
    const citiesTable = await getNode(connectionId, path);
    const primaryKey = await citiesTable.getChild('Keys/PK_Application_Cities');
    ```
    or
    ```typescript
    const path = ObjectExplorerNodePath
        .forTable('WideWorldImporters', 'Application', 'Cities')
        .join('Keys/PK_Application_Cities');
    const primaryKey = await getNode(connectionId, path);
    ```

File: src/sql/parts/registeredServer/common/objectExplorerService.ts (hunk 2 above)
Comment (MattIrv) [reply src/sql/parts/registeredServer/common/objectExplorerService.ts]: I refactored things a bit to use the VS Code expansion enum and make the rest of the code match up a bit more with their way of doing things. In the future we may want to refactor things a bit more if we end up exposing similar APIs for other trees, but that should be fairly straightforward now

File: src/sql/parts/registeredServer/common/objectExplorerService.ts (hunk 3 above)
Comment (MattIrv) [reply src/sql/parts/registeredServer/common/objectExplorerService.ts]: I'll be documenting all of it. In particular for a connection you might get it from a call to the connection APIs or from having an existing node (such as from `getActiveConnectionNodes`) and the next step of my work is to build an API that lets you get a node path from object metadata.

File: src/sql/parts/registeredServer/common/objectExplorerService.ts (hunk 8 above)
Comment (MattIrv) [reply src/sql/parts/registeredServer/common/objectExplorerService.ts]: Thanks for the async/await tip, it made this code much cleaner!

File: src/sql/sqlops.d.ts (hunk 12 above)
Comment (MattIrv) [reply src/sql/sqlops.d.ts]: Thanks for taking a look at this and for your feedback. I agree that building node paths as an API user isn't great. After this I'll start working on something like the `ObjectExplorerNodePath` factory you're suggesting

File: src/sql/parts/registeredServer/common/objectExplorerService.ts (hunk 8 above)
Comment (anthonydresser) [reply src/sql/parts/registeredServer/common/objectExplorerService.ts]: I would recommend not using async await, although it might make the code look pretty; because we are targeting an older version of javascript, it makes debugging nearly impossible. Just take a look at the output of a async await function.

Comment (MattIrv) [reply src/sql/parts/registeredServer/common/objectExplorerService.ts]: @anthonydresser can you explain what you mean? I did a lot of debugging while trying to get this code right and didn't run into any major problems, but maybe I just got lucky. In fact I almost felt that it was easier to debug since now I can just step over async calls if I want to

Comment (anthonydresser) [reply src/sql/parts/registeredServer/common/objectExplorerService.ts]: I found that when I tried to step into and through to await functions I would often be facing the underlying generator/await that is the javascript generated from using async/await, which forces you to debug very ugly javascript. Also, you have to be careful when it comes to errors, to properly get errors from await functions you have to wrap them in try/catches.

**Comment by MattIrv** [comment]
I added some basic docs about the extensibility APIs including sample code at https://github.com/Microsoft/sqlopsstudio/wiki/Extensibility-API. In the future we'll want to provide more thorough info, probably generated from the JSDoc in our code, but hopefully this helps for now!

File: src/sql/parts/registeredServer/common/objectExplorerService.ts [hunk 13]
```diff
@@ -267,24 +275,33 @@ export class ObjectExplorerService implements IObjectExplorerService {
...
-				self._sessions[session.sessionId].nodes[nodePath] = {
-					expandHandler: ((expandResult) => {
-						if (expandResult && !expandResult.errorMessage) {
-							resolve(expandResult);
-						}
-						else {
-							reject(expandResult ? expandResult.errorMessage : undefined);
-						}
+				let newRequest = false;
+				if (!self._sessions[session.sessionId].nodes[nodePath]) {
```
Comment (abist) [inline src/sql/parts/registeredServer/common/objectExplorerService.ts]: might want to check if `self._sessions[session.sessionId].nodes` exists as well

File: .travis.yml [hunk 14]
```diff
@@ -5,9 +5,7 @@ os:
   - linux
   - osx
 
-cache:
```
Comment (anthonydresser) [inline .travis.yml]: can we revert this change
